python/kvikio/tests/test_http_io.py

```diff
@@ -100,3 +156,53 @@ def test_no_range_support(http_server, tmpdir, xp):
             OverflowError, match="maybe the server doesn't support file ranges?"
         ):
             f.read(b, size=10, file_offset=10)
+
+
+def test_retry_http_500_ok(tmpdir, xp):
+    a = xp.arange(100, dtype="uint8")
+    a.tofile(tmpdir / "a")
+
+    with LocalHttpServer(
+        tmpdir,
+        max_lifetime=60,
+        handler=HTTP500Handler,
+        handler_options={"error_counter": ErrorCounter()},
+    ) as server:
+        http_server = server.url
+        b = xp.empty_like(a)
+        with kvikio.RemoteFile.open_http(f"{http_server}/a") as f:
+            assert f.nbytes() == a.nbytes
+            assert f"{http_server}/a" in str(f)
+            f.read(b)
+
+
+def test_retry_http_500_fails(tmpdir, xp):
+    with LocalHttpServer(
+        tmpdir,
+        max_lifetime=60,
+        handler=HTTP500Handler,
+        handler_options={"error_counter": ErrorCounter(), "max_error_count": 100},
+    ) as server:
+        a = xp.arange(100, dtype="uint8")
+        a.tofile(tmpdir / "a")
+        b = xp.empty_like(a)
+
+        with pytest.raises(RuntimeError, match="503"):
+            with kvikio.RemoteFile.open_http(f"{server.url}/a") as f:
+                f.read(b)
+
+
+def test_set_http_status_code(tmpdir, xp):
+    with LocalHttpServer(
+        tmpdir,
+        max_lifetime=60,
+        handler=HTTP500Handler,
+        handler_options={"error_counter": ErrorCounter()},
+    ) as server:
+        http_server = server.url
+        with kvikio.defaults.set_http_status_codes([429]):
+            # this raises on the first 503 error, since it's not in the list.
+            assert kvikio.defaults.http_status_codes() == [429]
+            with pytest.raises(RuntimeError, match="503"):
```

Comment:
As I start to think about #613 (timeouts) I'm wondering if kvikio should raise this a specialized exception type rather than a generic RuntimeError.

Doing a full exception hierarchy like [httpx](https://www.python-httpx.org/exceptions/#exception-classes) may be a bit much, but at least differentiating between a completed HTTP request with a 500 status code and a timeout is probably worthwhile.

Reply:
I think that would be a great idea. The current exception origination is very basic.

Reply:
Opened https://github.com/rapidsai/kvikio/issues/624 for that.